Support image resolution for init containers

skonto opened the pull request:
As per title, adds support for image resolution. Part of the work [here](https://github.com/knative/serving/issues/12024).
Introduces a new field in the revision status to keep the init container statuses.
Sample run with:
```
apiVersion: serving.knative.dev/v1
kind: Service
metadata:
  name: initcontainers
  namespace: default
spec:
  template:
    spec:
      containers:
        - imagePullPolicy: Always
          image: docker.io/skonto/emptydir
          volumeMounts:
            - name: data
              mountPath: /data
          env:
            - name: DATA_PATH
              value: /data
      initContainers:
        - imagePullPolicy: Always
          image: busybox
          name: init
          volumeMounts:
            - name: data
              mountPath: /data
      volumes:
        - name: data
          emptyDir: {}
```
![image](https://user-images.githubusercontent.com/7945591/139454381-8f921dc4-9967-4d8f-880e-681e8f3047e7.png)
![image](https://user-images.githubusercontent.com/7945591/139454415-61b21873-dab8-4b20-9765-2646c2073794.png)

skonto commented:
Test failures:
```
--- FAIL: TestSvcToSvcViaActivator (0.02s)
    stream.go:254: D 11:52:05.839 controller-695f6b9dc4-lqvf9 [knative.dev.serving.pkg.reconciler.service.Reconciler] [serving-tests/svc-to-svc-via-activator-b-disabled-yozlgojj] Adding to queue serving-tests/svc-to-svc-via-activator-b-disabled-yozlgojj (depth: 1)
    stream.go:254: D 11:52:05.895 controller-695f6b9dc4-lqvf9 [knative.dev.serving.pkg.reconciler.gc.reconciler] [serving-tests/svc-to-svc-via-activator-b-disabled-yozlgojj] Adding to queue serving-tests/svc-t

=== FAIL: test/scale TestScaleToN/scale-200/022-of-200 (399.27s)
    scale.go:103: Waiting for all to become ready
    stream.go:267: I 12:01:36.968 chaosduck-595dfbc79-zmmw4 chaosduck 2021/10/20 12:01:36 Got components: main.components{"autoscaler":sets.String{"autoscaler-679cf5796c-5dqmx":sets.Empty{}, "autoscaler-679cf5796
```

skonto commented on test/e2e/initcontainers/initcontainers_test.go:
```diff
@@ -90,4 +94,25 @@ func TestInitContainers(t *testing.T) {
+		if validDigest, err := shared.ValidateImageDigest(t, names.Image, initContainerStatus.ImageDigest); !validDigest {
+			return false, fmt.Errorf("imageDigest %s is not valid for imageName %s: %w", initContainerStatus.ImageDigest, initContainerStatus.Name, err)
+		}
+		if initContainerStatus.Name != "initsetup" {
```
This should be moved under `conformance` in the future (same as the sidecar containers status validation).

skonto commented on test/e2e/initcontainers/initcontainers_test.go:
```diff
@@ -90,4 +94,25 @@ func TestInitContainers(t *testing.T) {
+			return true, errors.New("init image digest resolution failed")
+		}
+		initContainerStatus := r.Status.InitContainerStatuses[0]
+		if validDigest, err := shared.ValidateImageDigest(t, names.Image, initContainerStatus.ImageDigest); !validDigest {
```
This should be moved under `conformance` in the future (same as the sidecar containers status validation).

skonto commented:
Failures not relevant:

```
--- FAIL: TestRouteCreation (32.26s)
    route_test.go:157: Creating a new Route and Configuration
    crd.go:36:  resource {<nil> <*>{&TypeMeta{Kind:,APIVersion:,} &ObjectMeta{Name:route-creation-
--- FAIL: TestSvcToSvcViaActivator (0.01s)
    stream.go:254: D 14:06:33.439 net-istio-controller-f46974458-wcxq5 [knative.dev.net-istio.pkg.reconciler.serverlessservice.reconciler] [serving-tests/svc-to-svc-via-activator-a-disabled-iknedqnz-00001] Adding to queue serving-tests/svc-to-svc-via-activator-a-disabled-iknedqnz-00001 (depth: 1)
    stream.go:254: I 14:06:33.439 net-istio-controller-f46974458-wcxq5 [knative.dev.net-istio.pkg.reconciler.serverlessservice.reconciler] [serving-tests/svc-to-svc-via-activator-a-disabled-iknedqnz-00001] 

```

skonto commented:
@julz @dprotaso gentle ping.

skonto commented:
@julz @dprotaso gentle ping.

julz commented on pkg/reconciler/revision/background.go:
```diff
@@ -157,38 +162,45 @@ func (r *backgroundResolver) Resolve(logger *zap.SugaredLogger, rev *v1.Revision
-	logger.Debugf("Resolve returned %d resolved images for revision", len(array))
-	return array, nil
+	logger.Debugf("Resolve returned %d resolved images for revision", len(statuses)+len(initContainerStatuses))
+	return statuses, initContainerStatuses, nil
```
this may just be me, but I feel like initContainers should probably come first (because they run before containers do)

skonto replied:
Yes makes sense ok.

julz commented on pkg/reconciler/revision/background.go:
```diff
@@ -206,6 +218,20 @@ func (r *backgroundResolver) addWorkItems(rev *v1.Revision, name types.Namespace
+			isInitContainer: true,
+		}
+
+		r.results[name].workItems[len(rev.Spec.Containers)+i] = item
```
might be better to use `append` here (and do `make([]workItem, 0, totalNumOfContainers)` above)

skonto replied:
ok let me check

julz commented on pkg/reconciler/revision/background.go:
```diff
@@ -64,8 +64,12 @@ type resolveResult struct {
-	err      error
+
+	// initContainerStatuses like `statuses` is a map of init container index to resolved status.
+	initContainerStatuses map[int]v1.ContainerStatus
```
an alternative approach might be to keep just one internal containerStatuses field, and build the two result arrays at the end of Resolve. That might simplify the logic a bit (we could drop isInitContainer and related if/else branches, I think). I'd be OK to land this as is and I can try that refactor in a follow-up, though.

julz replied:
I think it was just very slightly easier originally to do it with an int since that way you can build up the array and return it / I didn't think of using a string lookup when I wrote v1 of this :). I'll happily dig back in to this and refactor though once this lands (unless @skonto particularly wants to)

skonto replied:
I didnt want to create an array where I have a convention about ordering here. We agreed not to have it at the api level as it is simpler when statuses are reported in separate manner and my thought was that it is also better to keep it simple internally for readability and easy mapping of the values to the two status fields. I am not sure if there is a big gain here, but I will refactor no need to do it in a separate one.

dprotaso reviewed (commented):
so I'm good with this but I agree with @julz we should refactor some of this code in a separate PR

skonto commented:
> Any reason why this isn't just a map[string]string where the (k,v) is (image,resolved image with sha)?
I keep the index as with `result.statuses[item.index]` to show the status in order in the revision status field.
This goes back to the existing code.

skonto commented:
What is the rational behind checking the latency. Here is what I get for the failing unit test:

```
 logger.go:130: 2021-11-10T14:37:46.339Z	DEBUG	revision/background.go:234	Processing image "img2" from revision "ns/rev" (depth: 0)
    logger.go:130: 2021-11-10T14:37:46.339Z	DEBUG	revision/background.go:249	Resolving image "img2" from revision "ns/rev" to digest
    logger.go:130: 2021-11-10T14:37:46.339Z	DEBUG	revision/background.go:251	Resolved image "img2" from revision "ns/rev" to digest "", failed
    logger.go:130: 2021-11-10T14:37:46.339Z	DEBUG	revision/background.go:170	Resolve returned the resolved error: Unable to fetch image "img2": failed to resolve image to digest: failed
    background_test.go:281: latency = 1.225503ms, want at least 100ms
```
This is confusing as the test passes locally. Do I need to remove that part?

julz commented:
@skonto that test is validating that we correctly do exponential back-off (to avoid slamming registries and avoid DOS attacks) when image fetches fail. I think if it's passing locally this might be a false negative due to being slow for other reasons, but we need it to consistently take at least the given time (and the back-off should ensure that happens). Probably also the time is a bit too low if it's sometimes passing when it shouldn't so we might want to bump it higher.

skonto commented:
@julz since we enqueue the revision at each image failure should not the rateLimiter take care of that?

julz commented:
yeah it should, the test is there because at one point it didn't (due to items not comparing as equal because equality in go is annoying iirc).

skonto commented:
Ok that was tricky, it was failing with "-race". The test initially had an assumption of one image failing but I added two.
That caused revision resolution to finish earlier at some point (second iteration) as completionCallback would be called by 
some workItem that was left in the queue. That is why the fast finish time. It signaled the channel we have in the test too early (channel stays the same).

julz commented on pkg/reconciler/revision/background.go:
```diff
@@ -157,53 +157,71 @@ func (r *backgroundResolver) Resolve(logger *zap.SugaredLogger, rev *v1.Revision
+		r.results[name].imagesToBeResolved.Insert(container.Image)
+	}
+
+	for _, container := range append(rev.Spec.InitContainers, rev.Spec.Containers...) {
```
I'm probably missing something obvious here :), but why do we need to loop the same list twice?

skonto replied:
Because I wanted to initialize the static set with the image names before the work items are added to the queue.
It affects ready() check at least.

julz commented on pkg/reconciler/revision/background.go:
```diff
@@ -157,53 +157,71 @@ func (r *backgroundResolver) Resolve(logger *zap.SugaredLogger, rev *v1.Revision
 		item := workItem{
 			revision: name,
 			timeout:  timeout,
 			name:     container.Name,
```
I think workItem.name may be unused now since we're resolving by image id (so one image can be for multiple containers)

skonto replied:
Reply bellow as these two comments are related.

julz commented on pkg/reconciler/revision/background.go:
```diff
@@ -157,53 +157,71 @@ func (r *backgroundResolver) Resolve(logger *zap.SugaredLogger, rev *v1.Revision
-
-		r.results[name].workItems[i] = item
+		r.results[name].workItems = append(r.results[name].workItems, item)
 		r.queue.AddRateLimited(item)
```
if two containers use the same image, will we end up with two workItems both resolving that image, and if so is that intended/safe? (e.g. you could have one complete after the overall revision has already reported success/error). Could we only add one work item for each image maybe?

skonto replied:
That was the intention to keep the same semantics as before when adding items to the queue. Didnt want to break that.
Might be better to optimize it. Will test.

julz commented on pkg/reconciler/revision/revision.go:
```diff
@@ -93,16 +98,22 @@ func (c *Reconciler) reconcileDigest(ctx context.Context, rev *v1.Revision) (boo
-		rev.Status.ContainerStatuses = statuses
+
+	if len(statuses) > 0 || len(initContainerStatuses) > 0 {
+		if len(statuses) > 0 {
+			rev.Status.ContainerStatuses = statuses
+		}
+		if len(initContainerStatuses) > 0 {
+			rev.Status.InitContainerStatuses = initContainerStatuses
+		}
```
any reason we can't assign in either case? (given nil array and empty array should be the same when serialised I think, and we already checked that one of status/initStatus is non-nil so the resolve happened); i.e:

```suggestion
		rev.Status.ContainerStatuses = statuses
		rev.Status.InitContainerStatuses = initContainerStatuses
```

skonto replied:
In the past we had: 

```
	if len(statuses) > 0 {
		rev.Status.ContainerStatuses = statuses
		return true, nil
	}
```
So I wanted to keep the same semantics.

julz commented on pkg/reconciler/revision/revision.go:
```diff
@@ -71,12 +71,17 @@ var (
 
 func (c *Reconciler) reconcileDigest(ctx context.Context, rev *v1.Revision) (bool, error) {
+	totalNumOfContainers := len(rev.Spec.Containers) + len(rev.Spec.InitContainers)
 	if rev.Status.ContainerStatuses == nil {
 		rev.Status.ContainerStatuses = make([]v1.ContainerStatus, 0, len(rev.Spec.Containers))
 	}
+	if rev.Status.InitContainerStatuses == nil {
+		rev.Status.InitContainerStatuses = make([]v1.ContainerStatus, 0, len(rev.Spec.InitContainers))
+	}
```
I can't for the life of me figure out why we bother to initialise these arrays only to override them straight away 🤔 I know this is just doing what the pre-existing code does, but I'm almost sure we can drop the whole block

skonto replied:
Left over will omit.

julz reviewed (commented):
definitely think I prefer this way with the map[string]

skonto commented:
@julz I addressed the issues. I didnt see an issue with serialization null vs empty at least the `kubectl` level.
In addition, now one work item is added per image only.

julz commented on pkg/reconciler/revision/background.go:
```diff
@@ -157,54 +156,70 @@ func (r *backgroundResolver) Resolve(logger *zap.SugaredLogger, rev *v1.Revision
-			image:    container.Image,
-			index:    i,
+	for _, container := range append(rev.Spec.InitContainers, rev.Spec.Containers...) {
+		if !r.results[name].imagesToBeResolved.Has(container.Image) {
```
total nit, but I'd make this 
```suggestion
		if r.results[name].imagesToBeResolved.Has(container.Image) { continue }
```

and then outdent the main path

skonto replied:
Ok I can change.

julz commented on pkg/reconciler/revision/background.go:
```diff
@@ -157,54 +156,70 @@ func (r *backgroundResolver) Resolve(logger *zap.SugaredLogger, rev *v1.Revision
-		array[k] = v
+	initContainerStatuses = make([]v1.ContainerStatus, len(rev.Spec.InitContainers))
+	for i, container := range rev.Spec.InitContainers {
+		if resolvedDigest, ok := ret.imagesResolved[container.Image]; ok {
```
is it actually possible for imagesResolved not to have an entry here? If not I might drop the if/ok

julz commented on pkg/reconciler/revision/background.go:
```diff
@@ -157,54 +156,70 @@ func (r *backgroundResolver) Resolve(logger *zap.SugaredLogger, rev *v1.Revision
+				image:    container.Image,
+			}
+			r.results[name].workItems = append(r.results[name].workItems, item)
+			r.results[name].imagesToBeResolved.Insert(container.Image)
```
just to say this out loud: the reason this is safe and there's no race where we finish resolving an image and decide we're done before imagesToBeResolved is finished being set up is that both addWorkItems and processing results hold the same lock, right?

skonto replied:
Yes that was my rationale.

julz reviewed (commented):
lgtm bar a couple last-minute nits!

julz commented on pkg/reconciler/revision/background.go:
```diff
@@ -157,53 +156,69 @@ func (r *backgroundResolver) Resolve(logger *zap.SugaredLogger, rev *v1.Revision
-		array[k] = v
+	statuses = make([]v1.ContainerStatus, len(rev.Spec.Containers))
+	for i, container := range rev.Spec.Containers {
+		if resolvedDigest, ok := ret.imagesResolved[container.Image]; ok {
```
missed one, sorry

skonto commented:
Failures are irrelevant:

```
--- FAIL: TestSvcToSvcViaActivator (0.02s)
    stream.go:254: D 16:33:38.134 controller-9845d9875-p6jj7 [knative.dev.serving.pkg.reconciler.service.Reconciler] [serving-tests/svc-to-svc-via-activator-a-disabled-zutkyefx] Adding to queue serving-tests/svc-to-svc-via-activator-a-disabled-zutkyefx (depth: 1)
    stream.go:254: D 16:33:38.198 controller-9845d9875-p6jj7 [knative
   image_pull_error_test.go:73: When the images are not pulled, the revision should have error status.
    image_pull_error_test.go:86: Failed to validate revision state: the Revision image-pull-error-hjadoauf-00001 ReadyCondition = (Reason="Unschedulable", Message="0/2 nodes are available: 1 Insufficient cpu, 1 node(s) had taint {node-role.kubernetes.io/master: }, that the pod didn't tolerate."), wantReasons: [ImagePullBackOff ErrImagePull]
```

julz reviewed (approved):
thanks @skonto!

/lgtm
/approve
